File: 백준/Silver/2579. 계단 오르기/계단 오르기.swift
let n = Int(readLine()!)!
var stairs = Array(repeating: 0, count: n+1)
var d = Array(repeating: Array(repeating: 0, count: 3), 
              count: n+1)
for i in 1...n { stairs[i] = Int(readLine()!)! }

if(n == 1) { print(stairs[1]) }
else if(n == 2) { print(stairs[1] + stairs[2]) }
else {
    d[1][1] = stairs[1]
    d[1][2] = 0
    d[2][1] = stairs[2]
    d[2][2] = stairs[1] + stairs[2]
    
    for i in 3...n {
        d[i][1] = max(d[i-2][1], d[i-2][2]) + stairs[i]
        d[i][2] = d[i-1][1] + stairs[i]
    }
    print(max(d[n][1], d[n][2]))
}

File: 백준/Silver/2579. 계단 오르기/계단 오르기.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine()); // 계단 수
        int[] s = new int[n+1]; // 계단
        int[][] d = new int[n+1][3]; // 획득 점수
        for(int i=1; i<=n; i++) s[i] = Integer.parseInt(br.readLine());
        if(n == 1) {
            System.out.println(s[1]);
            return;
        }
        
        d[1][1] = s[1];
        d[1][2] = 0;
        d[2][1] = s[2];
        d[2][2] = s[1] + s[2];

        for(int i=3; i<=n; i++) {
            d[i][1] = Math.max(d[i-2][1], d[i-2][2])+s[i];
            d[i][2] = d[i-1][1] + s[i];
        }
        
        System.out.println(Math.max(d[n][1], d[n][2]));
    }
}